server.go

```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+	// We don't process anything else by them in this time so that we
+	// have an idea of when we should hear back from them - else the idle
+	// timeout could fire when we were only half done sending the blocks.
+	if numAdded > 0 {
+		<-doneChan
+	}
+}
+
+// handleGetBlocksMsg is invoked when a peer receives a getblocks bitcoin
+// message.
+func (s *server) handleGetBlocksMsg(p *peer.Peer, msg *wire.MsgGetBlocks) {
+	// Return all block hashes to the latest one (up to max per message) if
+	// no stop hash was specified.
+	// Attempt to find the ending index of the stop hash if specified.
+	endIdx := database.AllShas
+	if !msg.HashStop.IsEqual(&zeroHash) {
+		height, err := s.db.FetchBlockHeightBySha(&msg.HashStop)
+		if err == nil {
+			endIdx = height + 1
+		}
+	}
+
+	// Find the most recent known block based on the block locator.
+	// Use the block after the genesis block if no other blocks in the
+	// provided locator are known.  This does mean the client will start
+	// over with the genesis block if unknown block locators are provided.
+	// This mirrors the behavior in the reference implementation.
+	startIdx := int64(1)
+	for _, hash := range msg.BlockLocatorHashes {
+		height, err := s.db.FetchBlockHeightBySha(hash)
+		if err == nil {
+			// Start with the next hash since we know this one.
+			startIdx = height + 1
+			break
+		}
+	}
+
+	// Don't attempt to fetch more than we can put into a single message.
+	autoContinue := false
+	if endIdx-startIdx > wire.MaxBlocksPerMsg {
+		endIdx = startIdx + wire.MaxBlocksPerMsg
+		autoContinue = true
+	}
+
+	// Generate inventory message.
+	//
+	// The FetchBlockBySha call is limited to a maximum number of hashes
+	// per invocation.  Since the maximum number of inventory per message
+	// might be larger, call it multiple times with the appropriate indices
+	// as needed.
+	invMsg := wire.NewMsgInv()
+	for start := startIdx; start < endIdx; {
+		// Fetch the inventory from the block database.
+		hashList, err := s.db.FetchHeightRange(start, endIdx)
+		if err != nil {
+			peerLog.Warnf("Block lookup failed: %v", err)
+			return
+		}
+
+		// The database did not return any further hashes.  Break out of
+		// the loop now.
+		if len(hashList) == 0 {
+			break
+		}
+
+		// Add block inventory to the message.
+		for _, hash := range hashList {
+			hashCopy := hash
+			iv := wire.NewInvVect(wire.InvTypeBlock, &hashCopy)
+			invMsg.AddInvVect(iv)
+		}
+		start += int64(len(hashList))
+	}
+
+	// Send the inventory message if there is anything to send.
+	if len(invMsg.InvList) > 0 {
+		invListLen := len(invMsg.InvList)
+		if autoContinue && invListLen == wire.MaxBlocksPerMsg {
+			// Intentionally use a copy of the final hash so there
+			// is not a reference into the inventory slice which
+			// would prevent the entire slice from being eligible
+			// for GC as soon as it's sent.
+			s.blockManager.peers[p.ID()].continueHash = &invMsg.InvList[invListLen-1].Hash
```

Comment:
What happened to the hash copy?  As the comment states, it is needed to avoid delaying GC.